lsp/src/main/scala/org/ensime/lsp/rpc/companions.scala

```diff
@@ -0,0 +1,136 @@
+// Copyright: 2010 - 2017 https://github.com/ensime/ensime-server/graphs
+// License: http://www.gnu.org/licenses/gpl-3.0.en.html
+package org.ensime.lsp.rpc.companions
+
+import spray.json.{ JsObject, JsonFormat }
+import org.ensime.lsp.rpc.messages._
+import JsonRpcMessages._
+import org.ensime.lsp.rpc.JsInnerFormats.JsInnerField
+import shapeless.tag
+
+import scala.util.{ Failure, Success, Try }
+
+sealed trait RpcCompanionError {
+  val describe: String
+}
+
+case object UnknownMethod extends RpcCompanionError {
+  override val describe = "unknown method"
+}
+case object NoParams extends RpcCompanionError {
+  override val describe = "parameters must be given"
+}
+case object NoNamedParams extends RpcCompanionError {
+  override val describe = "named parameters must be given"
+}
+case class OtherError(err: String) extends RpcCompanionError {
+  override val describe = err
+}
+
+object RpcCompanionError {
+  def apply(err: String): OtherError = OtherError(err)
+}
+
+case class RpcCommand[A](method: String)(implicit val format: JsonFormat[A])
```

Comment:
taking the JsonFormat here is bad form... it is mixing your data model with the polymorphism. Best put the requirement on the thing doing the marshalling. e.g. you wouldn't want `Option` to take implicit evidence for everything its `A` would want to do.

Reply:
I added implicit val to `RpcCommand` to explicitly denote what `JsonFormat`s connected with corresponding methods. Format is generated only in `RpcCommand` definition, not when the writing/reading is performed.